crates/worker/src/partition/leadership/durability_tracker.rs

```diff
@@ -0,0 +1,242 @@
+        }
+    }
+
+    fn sanitize_durability_mode(
+        &mut self,
+        input_durability_mode: Option<DurabilityMode>,
+    ) -> DurabilityMode {
+        let configuration = Configuration::pinned();
+        let has_snapshot_repository = configuration.worker.snapshots.destination.is_some();
+        let is_cluster =
+            Metadata::with_current(|m| m.nodes_config_ref().iter_role(Role::Worker).count() > 1);
+
+        let require_snapshots_notice = |durability_mode| {
+            if !has_snapshot_repository && self.last_warning_at.elapsed() > WARN_PERIOD {
+                warn!(
+                    %durability_mode,
+                    "Detected cluster environment with no snapshot repository configured. \
+                    Automatic log trimming is disabled, please refer to \
+                    https://docs.restate.dev/operate/snapshots/ for more."
+                );
+            }
+        };
+
+        let risk_notice = |durability_mode| {
+            if !has_snapshot_repository && self.last_warning_at.elapsed() > WARN_PERIOD {
+                warn!(
+                    "No snapshot repository is configured. \
+                    Due to the configured durability mode '{durability_mode}', the cluster might \
+                    not be able to move partitions to other nodes for fail-over or rebalancing. \
+                    Please configure the cluster to use a shared snapshot repository. Please refer \
+                    to https://docs.restate.dev/operate/snapshots/ for more."
+                );
+            }
+        };
+
+        // ## Special cases:
+        // - A standalone node:
+        //   - no snapshot repository configured. Trim after local durability -> `ReplicaSetOnly`
+        //   - snapshot repository configured. Trim after a snapshot is taken + local durability -> `Balanced` || `SnapshotAndReplicaSet`
+        //
+        //   Simply, if snapshot is not configured, our options are:
+        //   - None
+        //   - ReplicaSetOnly
+        //
+        //   If snapshot is configured:
+        //   - None
+        //   - ReplicaSetOnly (not recommended, can't recover if cluster)
+        //   - SnapshotAndReplicaSet (not supported / disabled)
+        //   - SnapshotOnly
+        let durability_mode = input_durability_mode.unwrap_or({
+            // default depends on whether we have snapshot repository configured or not
+            if has_snapshot_repository {
+                DurabilityMode::Balanced
+            } else {
+                DurabilityMode::ReplicaSetOnly
+            }
+        });
+
+        match durability_mode {
+            DurabilityMode::None => {}
+            DurabilityMode::ReplicaSetOnly if !is_cluster => {}
+            DurabilityMode::ReplicaSetOnly => {
+                // todo: remove when we support ad-hoc snapshot sharing
+                risk_notice(durability_mode);
+                self.last_warning_at = Instant::now();
```

Comment:
By always updating the `last_warning_at` independent of whether the notice was printed or not and given frequent enough `sanitize_durability_mode` calls, wouldn't we never see these warnings?